File: CPP/2680_MaximumOR.cpp
/*
You are given a 0-indexed integer array nums of length n and an integer k. In an operation, you can choose an element and multiply it by 2.

Return the maximum possible value of nums[0] | nums[1] | ... | nums[n - 1] that can be obtained after applying the operation on nums at most k times.

Note that a | b denotes the bitwise or between two integers a and b.

 

Example 1:

Input: nums = [12,9], k = 1
Output: 30
Explanation: If we apply the operation to index 1, our new array nums will be equal to [12,18]. Thus, we return the bitwise or of 12 and 18, which is 30.
Example 2:

Input: nums = [8,1,2], k = 2
Output: 35
Explanation: If we apply the operation twice on index 0, we yield a new array of [32,1,2]. Thus, we return 32|1|2 = 35.
 

Constraints:

1 <= nums.length <= 105
1 <= nums[i] <= 109
1 <= k <= 15
 */

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
 public:
  int64_t maximumOr(const vector<int>& nums, int k) {
    int all_xor = 0;
    int same_bits = 0;
    for (int num : nums) {
      same_bits |= all_xor & num;
      all_xor |= num;
    }
    int64_t res = 0;
    for (int num : nums) {
      res = max(res, (all_xor ^ num) | same_bits | (int64_t) num << k);
    }
    return res;
  }
};

int main() {
  Solution sol;
  vector<int> nums{8, 1, 2};
  int k = 2;
  cout << sol.maximumOr(nums, k) << endl;
  vector<int> nums2{12, 9};
  int k2 = 1;
  cout << sol.maximumOr(nums2, k2) << endl;
}

File: 2025/2680_MaximumOR.py
from typing import List
from functools import reduce

class Solution: # wrong answer, the val << k may not in max_d_nums, example: # test 3
    def maximumOr(self, nums: List[int], k: int) -> int:
        digits_tuple = [(len(bin(num)) - 2, num) for num in nums]
        max_digits = max([d for d,_ in digits_tuple])
        max_d_nums = [v for d,v in digits_tuple if d == max_digits]
        non_max_d_nums = [v for d,v in digits_tuple if d < max_digits]
        non_max_flag = len(non_max_d_nums) > 0
        if non_max_flag:
            non_max_xor = reduce(lambda x,y: x^y, non_max_d_nums)
        res = 0
        for i,v in enumerate(max_d_nums):
            max_d_nums[i] = v << k
            curr_xor = reduce(lambda x,y: x^y, max_d_nums)
            if non_max_flag:
                res = max(res, curr_xor ^ non_max_xor)
            else:
                res = max(res, curr_xor)
            max_d_nums[i] = v
        return res
    
    
class Solution:
    def maximumOr(self, nums: List[int], k: int) -> int:
        all_xor = 0
        same_bits = 0
        for num in nums:
            same_bits |= all_xor & num
            all_xor |= num
        return max([(all_xor ^ num) | same_bits | num << k for num in nums])            

sol = Solution()
# test 1
nums = [12,9]
k = 1
print(sol.maximumOr(nums, k))
# test 2
nums = [8,1,2]
k = 2
print(sol.maximumOr(nums, k))
# test 3
nums = [98,54,6,34,66,63,52,39,62,46,75,28,65,18,37,18,97,13,80,33,69,91,78,19,40]
k = 2
print(sol.maximumOr(nums, k))
